### PR TITLE
feat: integrate interchain accounts

### PR DESCRIPTION
**Pull request by gsk967**
---
name: Pull Request
about: Submit a pull request
title: ''
labels: ''
assignees: '@boz @gosuri'

---

## Motivation and Context

- [ ] I have raised an issue to propose this change ([required](https://github.com/ovrclk/akash/blob/master/CONTRIBUTING.md))
- [ ] My issue has received approval from the maintainers or lead with the `design/approved` label

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I've read the [CONTRIBUTION](https://github.com/ovrclk/akash/blob/master/CONTRIBUTING.md) guide
- [ ] I have signed-off my commits with `git commit -s`
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: app/app.go
Comment (anilCSE) [inline app/app.go:112]:
```suggestion
```

**Review by anilCSE — APPROVED** [review]

**Comment by gsk967** [comment]
Simulations are failing due to issue from ibc-go , ibc-go does n't implement the simulation interfaces 
https://github.com/cosmos/ibc-go/issues/1352

File: app/app.go
Comment (boz) [inline app/app.go:562]: are upgrades ordered by name?  if so, can you make this `"01-ica-upgrade"` to leave room for `"02-foo"`?

File: x/icaauth/client/cli/flags.go
Comment (boz) [inline x/icaauth/client/cli/flags.go:19]: please don't use global variables for this - use the helpers like is done with other flags.

File: x/icaauth/client/cli/query.go
Comment (boz) [inline x/icaauth/client/cli/query.go:37]: why is `connectionId` a flag elsewhere but a param here?  let's make it consistent.

File: x/icaauth/ibc_module.go
Comment (boz) [inline x/icaauth/ibc_module.go:46]: looks like this should be debug level

File: x/icaauth/keeper/keeper.go
Comment (boz) [inline x/icaauth/keeper/keeper.go:25]: Make `Keeper` an interface, implemented by a package-local struct.

**Review by boz — CHANGES_REQUESTED** [review]
Looks great, thanks!

Please fix the lint failures (errors).

@troian any thoughts on when would be a good time to merge this?  I don't want it to impact the provider extraction.

File: app/app.go
Comment (anilCSE) [reply app/app.go]: No, it doesn't care about the order. It just cares about the upgrade height. But we can have a better upgrade name.

File: x/icaauth/client/cli/flags.go
Comment (gsk967) [reply x/icaauth/client/cli/flags.go]: ok

File: x/icaauth/ibc_module.go
Comment (gsk967) [reply x/icaauth/ibc_module.go]: ok

File: x/icaauth/keeper/keeper.go
Comment (gsk967) [reply x/icaauth/keeper/keeper.go:29]: ok

**Comment by boz** [comment]
> Simulations are failing due to issue from ibc-go , ibc-go does n't implement the simulation interfaces [cosmos/ibc-go#1352](https://github.com/cosmos/ibc-go/issues/1352)

@alessio and team provided us a fix for this, I will DM it to you.

**Comment by gsk967** [comment]
@boz  I fix the simulation issue, please can you look into changes

**Review by boz — COMMENTED** [review]
Thanks @anilCSE @gsk967!

We have a bit of release engineering to do before we merge this, will get it done soon!

@troian - defer to you on best way to handle this given with the repo split.

**Comment by troian** [comment]
@gsk967 we'll incorporate the change right after #1646

**Comment by gosuri** [comment]
@troian now that #1646 is merged, can we move forward with this one?

**Comment by troian** [comment]
@gosuri yeah, will merge it

**Comment by gsk967** [comment]
> @gsk967 mind rebasing pr

sure

**Review by troian — APPROVED** [review]

**Pull request merged**
